ACPs/194-streaming-asynchronous-execution/README.md

```diff
@@ -0,0 +1,394 @@
+
+$$
+\downarrow g_k \implies
+\begin{cases}
+    \downarrow \frac{g \cdot (p-1)}{p} \\
+    \uparrow \frac{s}{p}
+\end{cases}
+\implies \downarrow \Delta x_k
+\implies \downarrow M \cdot \exp\left(\frac{x_{i>k}}{K}\right)
+$$
+
+Given maximal gas consumption under (1), the monotonicity of $f$ implies (2).
+
+Since we are working with non-negative integers, it follows that multiplying a transaction's gas limit by the hypothetical gas price of (2) results in its worst-case gas cost.
+Any sender able to pay for this upper bound (in addition to value transfers) is guaranteed to be able to pay for the actual execution cost.
+Transaction _acceptance_ under worst-case cost validity is therefore a guarantee of _settlement_.
+
+### Queue DoS protection
+
+Worst-case cost validity only protects against DoS at the point of execution but leaves the queue vulnerable to high-limit, low-usage transactions.
+For example, a malicious user could send a transfer-only transaction (21k gas) with a limit set to multiple chunks.
+Although they would have to have sufficient funds to theoretically pay for all the chunks, they would never actually be charged this amount. Pushing a sufficient number of such transactions to the queue would artificially inflate the worst-case cost of other users.
+
+Gas limits are typically set higher than expected gas used to allow for a buffer should gas estimates be imprecise.
+A lower bound of 50% of the specified limit, for example, would allow for a 100% buffer on estimates without penalising the sender, while still disincentivising falsely high limits.
+
+#### Upper bound on queue DoS
+
+With $R$ (for rate) as already defined (gas capacity per second), we introduce $g_L$, $g_U$, and $g_C$ as gas _limit_, _used_ and _charged_ per transaction, respectively, and
+
+$$
+g_C := \max(g_U, \lambda \cdot g_L)
+$$
+
+where $\lambda$ enforces a lower bound on the charge.
+
+The actual gas excess $x_A$ has an upper bound of the worst-case excess $x_W$, both of which can be used  to calculate respective base fees $f_A$ and $f_W$ (the variable element of gas prices) from the existing exponential function:
+
+$$
+f := M \cdot \exp\left( \frac{x}{K} \right).
+$$
+
+Mallory is attempting to maximize the DoS ratio
+
+$$
+D := \frac{f_W}{f_A}
+$$
+
+by maximizing $\Sigma_{\forall i} (g_L - g_U)_i$ to maximize $x_W - x_A$.
+
+> [!TIP]
+> Although $D$ shadows a variable in ACP-176, that one is very different to anything here so there won't be confusion.
+
+Recall that the proportionality $p$ of capacity per second and target gas per second ($R = pT$) results in increasing excess such that
+
+$$
+x := x + \frac{g (p-1)}{p}.
+$$
+
+If we allow the queue to have a bandwidth of $R\lambda^{-1}$ gas/second and limit its length $\tau$ seconds or a back-pressure window of $\omega = \tau R \lambda^{-1}$ gas, we can derive an upper bound on the difference in the changes to worst-case and actual gas excess:
```

Comment:
We should comment explicitly that we are setting the maximum bandwidth of the queue to $\frac{R}{\lambda}$ to ensure that the queue rate limits never constrain the execution speed (while minimizing the bandwidth requirements).

Reply:
From the prior discussion:

Say we define:

```math
Capacity_{bandwidth}~:=~\dfrac{R}{\lambda}
```

This ensures that:

```math
g_C \geq R
```

Because (ignoring block packing inefficiencies):

```math
\begin{align}
g_C &= \sum_{\forall i}{\max\left({g_U}_i, \lambda \cdot {g_L}_i\right)}\\
g_C &\geq \sum_{\forall i}{\lambda \cdot {g_L}_i}\\
g_C &\geq \lambda \cdot \sum_{\forall i}{{g_L}_i}\\
g_C &\geq \lambda \cdot Capacity_{bandwidth}\\
g_C &\geq R
\end{align}
```